src/vibrato.rs

```diff
@@ -0,0 +1,143 @@
+// create a vibrato processor that is able to add vibrato effect to each block of audio data that works with multiple channels.
+
+use crate::ring_buffer::RingBuffer;
+
+use crate::lfo::LFO; 
+
+/// Vibrato is a struct that contains a delay line, an LFO, and a width, sample rate, and delay.
+pub struct Vibrato {
+    delay_line: Vec<RingBuffer<f32>>,
+    lfo: LFO,
+    width: f32,
+    sample_rate: f32,
+    delay: f32,
+}
+
+impl Vibrato {
+    /// Create a new vibrato processor with a given sample rate, maximum delay, delay, width, frequency, and number of channels.
+    pub fn new(sample_rate: f32, max_delay: f32, delay: f32, width: f32, frequency: f32, channels: usize) -> Vibrato {
+        // throw an error if width is bigger than max_delay
+        if width > max_delay {
+            panic!("Width is bigger than max_delay");
+        }
+        let mut delay_line = Vec::new();
+        let width = (width * sample_rate).round();
+        let lfo = LFO::new(sample_rate, frequency, 1.0);
+        let delay = (delay * sample_rate).round();
+        let max_delay = (max_delay * sample_rate).round();
+
+        let len_delay_line =delay + width * 2.0;
+        for _ in 0..channels {
+            let mut ring_buffer = RingBuffer::new((len_delay_line) as usize);
+            ring_buffer.set_read_index(0);
+            ring_buffer.set_write_index((len_delay_line-1.0) as usize);
+            delay_line.push(ring_buffer);
+        }
+
+
+        Vibrato {
+            delay_line,
+            lfo,
+            width,
+            sample_rate,
+            delay,
+        }
+    }
+
+    // process a block of audio data by adding vibrato effect to it
+    /// Process a block of audio data by adding vibrato effect to it.
+    pub fn process(&mut self, input: &mut [&mut [f32]], output: &mut [&mut [f32]]) {
+        for channel in 0..input.len() {
+            for sample in 0..input[channel].len() {
+                let delay = self.lfo.get_sample()*self.width + self.delay+1.0;
+                // dbg!(delay);
+                let read_index = self.delay_line[channel].get_read_index();
+                // dbg!(read_index);
+                // dbg!(self.delay_line[channel].peek());  
+                let write_index = self.delay_line[channel].get_write_index();
+                // dbg!(write_index);
+                let mut value = self.delay_line[channel].get_frac(delay);
+                // dbg!(value);
+
+                self.delay_line[channel].push(input[channel][sample]);
+                self.delay_line[channel].pop();
+                output[channel][sample] = value;
+                // dbg!(output[channel][sample]);
+                
+            }
+        }
+    }
+    /// Set the parameters of the vibrato processor.
+    pub fn set_param(&mut self, sample_rate: f32, delay: f32, width: f32, frequency: f32) {
```

Comment:
It might be better to call this set_params, that way it's clear that you're setting all of them at once. Also, you probably don't have to have sample_rate as a parameter because it's likely not going to change while the effect is processing.

Reply:
fixed as suggested